{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build P9 (ATPase) protein tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def extract_faa(gb_file, name):\n",
    "    i = 1\n",
    "    \n",
    "    faa_records = []\n",
    "    \n",
    "    for record in SeqIO.parse(gb_file, 'genbank'):\n",
    "        for feature in record.features:\n",
    "            if feature.type == 'CDS':\n",
    "                aa_seq = feature.qualifiers['translation'][0]\n",
    "                aa_record = SeqRecord(Seq(aa_seq), id=f'{name}_orf{i}', name='', description='')\n",
    "                faa_records.append(aa_record)\n",
    "                i += 1\n",
    "    \n",
    "    return faa_records\n",
    "\n",
    "def read_tblout(tblout):\n",
    "    df = pd.read_csv(tblout,\n",
    "                 delim_whitespace=True,\n",
    "                 comment='#',\n",
    "                 header=None)\n",
    "    \n",
    "    # parse and label column\n",
    "    df = df[df.columns[:5]]\n",
    "    df.columns = ['target', '-', 'query', '-',  'full_eval']\n",
    "    \n",
    "    \n",
    "    df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n",
    "    df = df[['acc', 'target', 'query', 'full_eval']]\n",
    "    \n",
    "    df = df[df['full_eval'] < 0.01]\n",
    "    \n",
    "    # select best hit for annotation\n",
    "    df = df.sort_values(by=['acc', 'target', 'full_eval'], ascending=[True, True, False])\n",
    "    df = df.drop_duplicates(subset=['target'], keep='first')\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get P9 sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# list with aa sequences for P9 proteins\n",
    "p9_list = []\n",
    "\n",
    "# result will be saved here\n",
    "p9_all_faa = '../data/tree/P9/P9.faa'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This study's"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get selection of alphatectiviruses\n",
    "sspp = ['PRDobsidian', 'PRDpistachio', 'PRDaquamarine',\n",
    "        'PRDcanary', 'PRDsepia', 'PRDwine',\n",
    "        'PRDavocado', 'PRDamber', 'PRDforest',\n",
    "        'PRDcobalt', 'PRDindigo']\n",
    "\n",
    "\n",
    "# append aminoacid sequences corresponding to those genomes\n",
    "for record in SeqIO.parse('../data/models/faa/IX.faa', 'fasta'):\n",
    "    isolate_name = record.id.split('_IX')[0]\n",
    "    if isolate_name in sspp:\n",
    "        record.id = isolate_name\n",
    "        record.name = ''\n",
    "        record.description = ''\n",
    "        p9_list.append(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run hmmer with ATPase model on ncbi tectiviruses\n",
    "\n",
    "#!hmmsearch \\\n",
    "#    --tblout '../data/tecti_genomes/NCBI/hmmer/p9.tbl' \\\n",
    "#    '../data/models/hmm/IX.2.hmm' \\\n",
    "#    '../data/tecti_genomes/NCBI/hmmer/all_ncbi_orfs.faa' \\\n",
    "#    > /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    }
   ],
   "source": [
    "# add p9 hits to list\n",
    "df_ncbi_hits = read_tblout('../data/tecti_genomes/NCBI/hmmer/p9.tbl')\n",
    "\n",
    "for record in SeqIO.parse('../data/tecti_genomes/NCBI/hmmer/all_ncbi_orfs.faa', 'fasta'):\n",
    "    if record.id in df_ncbi_hits[df_ncbi_hits['query'] == 'IX.2']['target'].to_list():\n",
    "        record.name = ''\n",
    "        record.description = ''\n",
    "        p9_list.append(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add PRD1 and PR4\n",
    "for record in SeqIO.parse('../data/tecti_genomes/NCBI/gb/ref_p9.faa', 'fasta'):\n",
    "    p9_list.append(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### JGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run hmmer with ATPase model on jgi imgvr tectiviruses\n",
    "\n",
    "#!hmmsearch \\\n",
    "#    --tblout '../data/tecti_genomes/JGI_IMGVR/hmmer/p9.tbl' \\\n",
    "#    '../data/models/hmm/IX.2.hmm' \\\n",
    "#    '../data/tecti_genomes/JGI_IMGVR/hmmer/all_jgi_orfs.faa' \\\n",
    "#    > /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    }
   ],
   "source": [
    "# add p9 hits to list\n",
    "df_jgi_hits = read_tblout('../data/tecti_genomes/JGI_IMGVR/hmmer/p9.tbl')\n",
    "\n",
    "for record in SeqIO.parse('../data/tecti_genomes/JGI_IMGVR/hmmer/all_jgi_orfs.faa', 'fasta'):\n",
    "    if record.id in df_jgi_hits[df_jgi_hits['query'] == 'IX.2']['target'].to_list():\n",
    "        record.name = ''\n",
    "        record.description = ''\n",
    "        p9_list.append(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Yutin (2018)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run hmmer with ATPase model on Yutin, et al. tectiviruses\n",
    "\n",
    "#!hmmsearch \\\n",
    "#    --tblout '../data/tecti_genomes/Yutin/hmmer/p9.tbl' \\\n",
    "#    '../data/models/hmm/IX.2.hmm' \\\n",
    "#    '../data/tecti_genomes/Yutin/hmmer/all_yutin_orfs.faa' \\\n",
    "#    > /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    }
   ],
   "source": [
    "# add p9 hits to list\n",
    "df_yutin_hits = read_tblout('../data/tecti_genomes/Yutin/hmmer/p9.tbl')\n",
    "\n",
    "for record in SeqIO.parse('../data/tecti_genomes/Yutin/hmmer/all_yutin_orfs.faa', 'fasta'):\n",
    "    if record.id in df_yutin_hits[df_yutin_hits['query'] == 'IX.2']['target'].to_list():\n",
    "        record.name = ''\n",
    "        record.description = ''\n",
    "        p9_list.append(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save result\n",
    "#SeqIO.write(p9_list, p9_all_faa, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Explore results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NCBI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>genus</th>\n",
       "      <th>species_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>len</th>\n",
       "      <th>host_class</th>\n",
       "      <th>host_sp</th>\n",
       "      <th>IX.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_011523.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus AP50</td>\n",
       "      <td>Bacillus phage AP50</td>\n",
       "      <td>14398</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "      <td>2.500000e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_005258.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus Bam35</td>\n",
       "      <td>Bacillus phage Bam35c</td>\n",
       "      <td>14935</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus thuringiensis</td>\n",
       "      <td>3.100000e-58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_006945.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus GIL16</td>\n",
       "      <td>Bacillus phage GIL16c</td>\n",
       "      <td>14844</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus thuringiensis</td>\n",
       "      <td>9.600000e-57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MZ089978.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus sato</td>\n",
       "      <td>Bacillus phage Sato</td>\n",
       "      <td>14852</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus cereus</td>\n",
       "      <td>1.500000e-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MZ089979.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus sole</td>\n",
       "      <td>Bacillus phage Sole</td>\n",
       "      <td>14444</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus cereus VD166</td>\n",
       "      <td>6.900000e-57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NC_022094.1</td>\n",
       "      <td>Betatectivirus</td>\n",
       "      <td>Betatectivirus Wip1</td>\n",
       "      <td>Bacillus phage Wip1</td>\n",
       "      <td>14319</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Bacillus anthracis</td>\n",
       "      <td>3.900000e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_055059.1</td>\n",
       "      <td>Deltatectivirus</td>\n",
       "      <td>Deltatectivirus forthebois</td>\n",
       "      <td>Streptomyces phage Forthebois</td>\n",
       "      <td>18251</td>\n",
       "      <td>Actinomycetia</td>\n",
       "      <td>Streptomyces scabiei</td>\n",
       "      <td>2.800000e-36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_055060.1</td>\n",
       "      <td>Deltatectivirus</td>\n",
       "      <td>Deltatectivirus wheeheim</td>\n",
       "      <td>Streptomyces phage WheeHeim</td>\n",
       "      <td>18266</td>\n",
       "      <td>Actinomycetia</td>\n",
       "      <td>Streptomyces scabiei</td>\n",
       "      <td>9.300000e-36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_055061.1</td>\n",
       "      <td>Epsilontectivirus</td>\n",
       "      <td>Epsilontectivirus toil</td>\n",
       "      <td>Rhodococcus phage Toil</td>\n",
       "      <td>17253</td>\n",
       "      <td>Actinomycetia</td>\n",
       "      <td>Rhodococcus opacus</td>\n",
       "      <td>1.900000e-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_042083.1</td>\n",
       "      <td>Gammatectivirus</td>\n",
       "      <td>Gammatectivirus GC1</td>\n",
       "      <td>Gluconobacter phage GC1</td>\n",
       "      <td>16523</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Gluconobacter cerinus</td>\n",
       "      <td>6.200000e-74</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc              genus                species_name  \\\n",
       "0  NC_011523.1     Betatectivirus         Betatectivirus AP50   \n",
       "1  NC_005258.1     Betatectivirus        Betatectivirus Bam35   \n",
       "2  NC_006945.1     Betatectivirus        Betatectivirus GIL16   \n",
       "3   MZ089978.1     Betatectivirus         Betatectivirus sato   \n",
       "4   MZ089979.1     Betatectivirus         Betatectivirus sole   \n",
       "5  NC_022094.1     Betatectivirus         Betatectivirus Wip1   \n",
       "6  NC_055059.1    Deltatectivirus  Deltatectivirus forthebois   \n",
       "7  NC_055060.1    Deltatectivirus    Deltatectivirus wheeheim   \n",
       "8  NC_055061.1  Epsilontectivirus      Epsilontectivirus toil   \n",
       "9  NC_042083.1    Gammatectivirus         Gammatectivirus GC1   \n",
       "\n",
       "                         isolate    len           host_class  \\\n",
       "0            Bacillus phage AP50  14398              Bacilli   \n",
       "1          Bacillus phage Bam35c  14935              Bacilli   \n",
       "2          Bacillus phage GIL16c  14844              Bacilli   \n",
       "3            Bacillus phage Sato  14852              Bacilli   \n",
       "4            Bacillus phage Sole  14444              Bacilli   \n",
       "5            Bacillus phage Wip1  14319              Bacilli   \n",
       "6  Streptomyces phage Forthebois  18251        Actinomycetia   \n",
       "7    Streptomyces phage WheeHeim  18266        Actinomycetia   \n",
       "8         Rhodococcus phage Toil  17253        Actinomycetia   \n",
       "9        Gluconobacter phage GC1  16523  Alphaproteobacteria   \n",
       "\n",
       "                  host_sp          IX.2  \n",
       "0      Bacillus anthracis  2.500000e-55  \n",
       "1  Bacillus thuringiensis  3.100000e-58  \n",
       "2  Bacillus thuringiensis  9.600000e-57  \n",
       "3         Bacillus cereus  1.500000e-56  \n",
       "4   Bacillus cereus VD166  6.900000e-57  \n",
       "5      Bacillus anthracis  3.900000e-55  \n",
       "6    Streptomyces scabiei  2.800000e-36  \n",
       "7    Streptomyces scabiei  9.300000e-36  \n",
       "8      Rhodococcus opacus  1.900000e-65  \n",
       "9   Gluconobacter cerinus  6.200000e-74  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read hits table\n",
    "df_ncbi_hits = read_tblout('../data/tecti_genomes/NCBI/hmmer/p9.tbl')\n",
    "\n",
    "# read metadata table\n",
    "df_ncbi_metadata = pd.read_csv('../data/tecti_genomes/NCBI/tectivirus_metadata.tsv', sep='\\t')\n",
    "df_ncbi_metadata = df_ncbi_metadata[df_ncbi_metadata['genus'] != '-']\n",
    "df_ncbi_metadata = df_ncbi_metadata[df_ncbi_metadata['genus'] != 'Alphatectivirus']\n",
    "\n",
    "pivot_ncbi_hits = df_ncbi_hits.pivot(index='acc', columns='query', values='full_eval').reset_index().fillna('-')\n",
    "df_ncbi_metadata.merge(pivot_ncbi_hits, on='acc', how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### JGI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>img_taxon_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>len</th>\n",
       "      <th>loc</th>\n",
       "      <th>IX.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ga0500005_002616</td>\n",
       "      <td>3300050116</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>19730</td>\n",
       "      <td>sweden</td>\n",
       "      <td>4.000000e-48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ga0500006_002682</td>\n",
       "      <td>3300050117</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>20094</td>\n",
       "      <td>sweden</td>\n",
       "      <td>4.000000e-48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ga0500006_012856</td>\n",
       "      <td>3300050117</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>5558</td>\n",
       "      <td>sweden</td>\n",
       "      <td>1.700000e-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ga0500009_001431</td>\n",
       "      <td>3300050120</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>17036</td>\n",
       "      <td>sweden</td>\n",
       "      <td>7.200000e-59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ga0500011_004634</td>\n",
       "      <td>3300050122</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>4396</td>\n",
       "      <td>sweden</td>\n",
       "      <td>3.400000e-62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ga0500012_004241</td>\n",
       "      <td>3300050123</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>4422</td>\n",
       "      <td>sweden</td>\n",
       "      <td>3.300000e-62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ga0500013_005094</td>\n",
       "      <td>3300050124</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>4816</td>\n",
       "      <td>sweden</td>\n",
       "      <td>3.300000e-62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ga0500017_0003996</td>\n",
       "      <td>3300050128</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>11142</td>\n",
       "      <td>sweden</td>\n",
       "      <td>4.000000e-48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ga0500018_001608</td>\n",
       "      <td>3300050129</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>18742</td>\n",
       "      <td>sweden</td>\n",
       "      <td>4.000000e-48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ga0500022_002121</td>\n",
       "      <td>3300050132</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>18125</td>\n",
       "      <td>sweden</td>\n",
       "      <td>9.800000e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Ga0500022_006983</td>\n",
       "      <td>3300050132</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>7686</td>\n",
       "      <td>sweden</td>\n",
       "      <td>9.400000e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Ga0500032_0002123</td>\n",
       "      <td>3300050138</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>19102</td>\n",
       "      <td>sweden</td>\n",
       "      <td>2.600000e-64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ga0501157_006299</td>\n",
       "      <td>3300050150</td>\n",
       "      <td>Alder tree leaf litter microbial communities f...</td>\n",
       "      <td>6962</td>\n",
       "      <td>washington</td>\n",
       "      <td>4.500000e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Ga0500030_003738</td>\n",
       "      <td>3300050192</td>\n",
       "      <td>Peatland microbial communities from Stordalen ...</td>\n",
       "      <td>16031</td>\n",
       "      <td>sweden</td>\n",
       "      <td>4.000000e-48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Ga0505706_00067931</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>17295</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.100000e-57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Ga0505706_00072834</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>16739</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.400000e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ga0505706_00075008</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>16516</td>\n",
       "      <td>harvard</td>\n",
       "      <td>3.200000e-40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ga0505706_00075641</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>16449</td>\n",
       "      <td>harvard</td>\n",
       "      <td>6.400000e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ga0505706_00078677</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>16158</td>\n",
       "      <td>harvard</td>\n",
       "      <td>2.900000e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ga0505706_00084197</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>15661</td>\n",
       "      <td>harvard</td>\n",
       "      <td>4.500000e-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Ga0505706_00084834</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>15610</td>\n",
       "      <td>harvard</td>\n",
       "      <td>3.700000e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Ga0505706_00087319</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>15398</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.500000e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Ga0505706_00093977</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>14878</td>\n",
       "      <td>harvard</td>\n",
       "      <td>2.900000e-52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ga0505706_00098891</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>14526</td>\n",
       "      <td>harvard</td>\n",
       "      <td>9.200000e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ga0505706_00117367</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>13401</td>\n",
       "      <td>harvard</td>\n",
       "      <td>7.000000e-58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Ga0505706_00127399</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>12896</td>\n",
       "      <td>harvard</td>\n",
       "      <td>5.600000e-51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Ga0505706_00137823</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>12426</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.300000e-41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Ga0505706_00138351</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>12402</td>\n",
       "      <td>harvard</td>\n",
       "      <td>4.500000e-59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Ga0505706_00153685</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>11796</td>\n",
       "      <td>harvard</td>\n",
       "      <td>9.500000e-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Ga0505706_00231171</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>9693</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.700000e-49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Ga0505706_00332338</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>8116</td>\n",
       "      <td>harvard</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Ga0505706_00334178</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>8094</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.200000e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Ga0505706_00344671</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>7972</td>\n",
       "      <td>harvard</td>\n",
       "      <td>6.000000e-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Ga0505706_00373652</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>7662</td>\n",
       "      <td>harvard</td>\n",
       "      <td>8.900000e-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Ga0505706_00425488</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>7184</td>\n",
       "      <td>harvard</td>\n",
       "      <td>1.600000e-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Ga0505706_00937364</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>4822</td>\n",
       "      <td>harvard</td>\n",
       "      <td>3.200000e-59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Ga0505706_00969868</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>4737</td>\n",
       "      <td>harvard</td>\n",
       "      <td>2.900000e-50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Ga0505706_01001969</td>\n",
       "      <td>3300050821</td>\n",
       "      <td>Co-assembly of Harvard Forest metagenomes from...</td>\n",
       "      <td>4659</td>\n",
       "      <td>harvard</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   acc  img_taxon_id  \\\n",
       "0     Ga0500005_002616    3300050116   \n",
       "1     Ga0500006_002682    3300050117   \n",
       "2     Ga0500006_012856    3300050117   \n",
       "3     Ga0500009_001431    3300050120   \n",
       "4     Ga0500011_004634    3300050122   \n",
       "5     Ga0500012_004241    3300050123   \n",
       "6     Ga0500013_005094    3300050124   \n",
       "7    Ga0500017_0003996    3300050128   \n",
       "8     Ga0500018_001608    3300050129   \n",
       "9     Ga0500022_002121    3300050132   \n",
       "10    Ga0500022_006983    3300050132   \n",
       "11   Ga0500032_0002123    3300050138   \n",
       "12    Ga0501157_006299    3300050150   \n",
       "13    Ga0500030_003738    3300050192   \n",
       "14  Ga0505706_00067931    3300050821   \n",
       "15  Ga0505706_00072834    3300050821   \n",
       "16  Ga0505706_00075008    3300050821   \n",
       "17  Ga0505706_00075641    3300050821   \n",
       "18  Ga0505706_00078677    3300050821   \n",
       "19  Ga0505706_00084197    3300050821   \n",
       "20  Ga0505706_00084834    3300050821   \n",
       "21  Ga0505706_00087319    3300050821   \n",
       "22  Ga0505706_00093977    3300050821   \n",
       "23  Ga0505706_00098891    3300050821   \n",
       "24  Ga0505706_00117367    3300050821   \n",
       "25  Ga0505706_00127399    3300050821   \n",
       "26  Ga0505706_00137823    3300050821   \n",
       "27  Ga0505706_00138351    3300050821   \n",
       "28  Ga0505706_00153685    3300050821   \n",
       "29  Ga0505706_00231171    3300050821   \n",
       "30  Ga0505706_00332338    3300050821   \n",
       "31  Ga0505706_00334178    3300050821   \n",
       "32  Ga0505706_00344671    3300050821   \n",
       "33  Ga0505706_00373652    3300050821   \n",
       "34  Ga0505706_00425488    3300050821   \n",
       "35  Ga0505706_00937364    3300050821   \n",
       "36  Ga0505706_00969868    3300050821   \n",
       "37  Ga0505706_01001969    3300050821   \n",
       "\n",
       "                                          sample_name    len         loc  \\\n",
       "0   Peatland microbial communities from Stordalen ...  19730      sweden   \n",
       "1   Peatland microbial communities from Stordalen ...  20094      sweden   \n",
       "2   Peatland microbial communities from Stordalen ...   5558      sweden   \n",
       "3   Peatland microbial communities from Stordalen ...  17036      sweden   \n",
       "4   Peatland microbial communities from Stordalen ...   4396      sweden   \n",
       "5   Peatland microbial communities from Stordalen ...   4422      sweden   \n",
       "6   Peatland microbial communities from Stordalen ...   4816      sweden   \n",
       "7   Peatland microbial communities from Stordalen ...  11142      sweden   \n",
       "8   Peatland microbial communities from Stordalen ...  18742      sweden   \n",
       "9   Peatland microbial communities from Stordalen ...  18125      sweden   \n",
       "10  Peatland microbial communities from Stordalen ...   7686      sweden   \n",
       "11  Peatland microbial communities from Stordalen ...  19102      sweden   \n",
       "12  Alder tree leaf litter microbial communities f...   6962  washington   \n",
       "13  Peatland microbial communities from Stordalen ...  16031      sweden   \n",
       "14  Co-assembly of Harvard Forest metagenomes from...  17295     harvard   \n",
       "15  Co-assembly of Harvard Forest metagenomes from...  16739     harvard   \n",
       "16  Co-assembly of Harvard Forest metagenomes from...  16516     harvard   \n",
       "17  Co-assembly of Harvard Forest metagenomes from...  16449     harvard   \n",
       "18  Co-assembly of Harvard Forest metagenomes from...  16158     harvard   \n",
       "19  Co-assembly of Harvard Forest metagenomes from...  15661     harvard   \n",
       "20  Co-assembly of Harvard Forest metagenomes from...  15610     harvard   \n",
       "21  Co-assembly of Harvard Forest metagenomes from...  15398     harvard   \n",
       "22  Co-assembly of Harvard Forest metagenomes from...  14878     harvard   \n",
       "23  Co-assembly of Harvard Forest metagenomes from...  14526     harvard   \n",
       "24  Co-assembly of Harvard Forest metagenomes from...  13401     harvard   \n",
       "25  Co-assembly of Harvard Forest metagenomes from...  12896     harvard   \n",
       "26  Co-assembly of Harvard Forest metagenomes from...  12426     harvard   \n",
       "27  Co-assembly of Harvard Forest metagenomes from...  12402     harvard   \n",
       "28  Co-assembly of Harvard Forest metagenomes from...  11796     harvard   \n",
       "29  Co-assembly of Harvard Forest metagenomes from...   9693     harvard   \n",
       "30  Co-assembly of Harvard Forest metagenomes from...   8116     harvard   \n",
       "31  Co-assembly of Harvard Forest metagenomes from...   8094     harvard   \n",
       "32  Co-assembly of Harvard Forest metagenomes from...   7972     harvard   \n",
       "33  Co-assembly of Harvard Forest metagenomes from...   7662     harvard   \n",
       "34  Co-assembly of Harvard Forest metagenomes from...   7184     harvard   \n",
       "35  Co-assembly of Harvard Forest metagenomes from...   4822     harvard   \n",
       "36  Co-assembly of Harvard Forest metagenomes from...   4737     harvard   \n",
       "37  Co-assembly of Harvard Forest metagenomes from...   4659     harvard   \n",
       "\n",
       "            IX.2  \n",
       "0   4.000000e-48  \n",
       "1   4.000000e-48  \n",
       "2   1.700000e-45  \n",
       "3   7.200000e-59  \n",
       "4   3.400000e-62  \n",
       "5   3.300000e-62  \n",
       "6   3.300000e-62  \n",
       "7   4.000000e-48  \n",
       "8   4.000000e-48  \n",
       "9   9.800000e-41  \n",
       "10  9.400000e-43  \n",
       "11  2.600000e-64  \n",
       "12  4.500000e-52  \n",
       "13  4.000000e-48  \n",
       "14  1.100000e-57  \n",
       "15  1.400000e-44  \n",
       "16  3.200000e-40  \n",
       "17  6.400000e-29  \n",
       "18  2.900000e-60  \n",
       "19  4.500000e-49  \n",
       "20  3.700000e-44  \n",
       "21  1.500000e-55  \n",
       "22  2.900000e-52  \n",
       "23  9.200000e-46  \n",
       "24  7.000000e-58  \n",
       "25  5.600000e-51  \n",
       "26  1.300000e-41  \n",
       "27  4.500000e-59  \n",
       "28  9.500000e-42  \n",
       "29  1.700000e-49  \n",
       "30           NaN  \n",
       "31  1.200000e-60  \n",
       "32  6.000000e-42  \n",
       "33  8.900000e-56  \n",
       "34  1.600000e-28  \n",
       "35  3.200000e-59  \n",
       "36  2.900000e-50  \n",
       "37           NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_jgi_hits = read_tblout('../data/tecti_genomes/JGI_IMGVR/hmmer/p9.tbl')\n",
    "\n",
    "df_jgi_metadata = pd.read_csv('../data/tecti_genomes/JGI_IMGVR/JGI_metadata.tsv', sep='\\t')\n",
    "\n",
    "pivot_jgi_hits = df_jgi_hits.pivot(index='acc', columns='query', values='full_eval').reset_index().fillna('-')\n",
    "df_jgi_metadata.merge(pivot_jgi_hits, on='acc', how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Yutin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31346/3433487581.py:27: FutureWarning: In a future version of pandas all arguments of StringMethods.rsplit except for the argument 'pat' will be keyword-only.\n",
      "  df['acc'] = df['target'].str.rsplit('_', 1, expand=True)[0]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>environment</th>\n",
       "      <th>len</th>\n",
       "      <th>IX.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LNFM01013825.1</td>\n",
       "      <td>Activated carbon metagenome</td>\n",
       "      <td>10153</td>\n",
       "      <td>4.000000e-60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FRDC01003407.1</td>\n",
       "      <td>freshwater metagenome</td>\n",
       "      <td>17409</td>\n",
       "      <td>3.200000e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LNFM01009513.1</td>\n",
       "      <td>Activated carbon metagenome</td>\n",
       "      <td>14103</td>\n",
       "      <td>1.900000e-32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JRYJ01001167.1</td>\n",
       "      <td>Activated sludge metagenome</td>\n",
       "      <td>17872</td>\n",
       "      <td>5.600000e-29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              acc                   environment    len          IX.2\n",
       "0  LNFM01013825.1   Activated carbon metagenome  10153  4.000000e-60\n",
       "1  FRDC01003407.1         freshwater metagenome  17409  3.200000e-25\n",
       "2  LNFM01009513.1   Activated carbon metagenome  14103  1.900000e-32\n",
       "3  JRYJ01001167.1  Activated sludge metagenome   17872  5.600000e-29"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_yutin_hits = read_tblout('../data/tecti_genomes/Yutin/hmmer/p9.tbl')\n",
    "\n",
    "# explore results\n",
    "df_yutin_metadata = pd.read_csv('../data/tecti_genomes/Yutin/yutin_metadata.tsv', sep='\\t')\n",
    "\n",
    "pivot_yutin_hits = df_yutin_hits.pivot(index='acc', columns='query', values='full_eval').reset_index().fillna('-')\n",
    "df_yutin_metadata.merge(pivot_yutin_hits, on='acc', how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# align resulting sequences to model\n",
    "#!hmmalign --trim '../data/models/hmm/IX.2.hmm' '../data/tree/P9/P9.faa' | esl-reformat --gapsym='-' afa - > '../data/tree/P9/P9.afa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# give them nice readable names for phylip format\n",
    "name_dict = dict(zip(df_ncbi_metadata['acc'], df_ncbi_metadata['species_name'].str.replace('tectivirus', '').str.replace(' ', '-')))\n",
    "name_dict.update(zip(df_jgi_metadata['acc'], df_jgi_metadata['loc'].str[:2] + df_jgi_metadata['acc'].str.split('_', expand=True)[1].str.lstrip('0')))\n",
    "name_dict.update(zip(df_yutin_metadata['acc'], df_yutin_metadata['acc'].str.split('.', expand=True)[0]))\n",
    "\n",
    "renamed = []\n",
    "for record in SeqIO.parse('../data/tree/P9/P9.afa', 'fasta'):\n",
    "    name = record.id.split('_orf')[0]\n",
    "    if name in name_dict.keys():\n",
    "        new_name = name_dict[name]\n",
    "    else:\n",
    "        new_name = name\n",
    "        \n",
    "    record.id = new_name\n",
    "    \n",
    "    renamed.append(record)\n",
    "    \n",
    "    \n",
    "#SeqIO.write(renamed, '../data/tree/P9/P9.phy', 'phylip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "aln_phy = '../data/tree/P9/P9.phy'\n",
    "phy_cmd = 'phyml '\\\n",
    "            '-d aa '\\\n",
    "            '-m LG ' \\\n",
    "            '-b -4 '\\\n",
    "            '-v 0.0 '\\\n",
    "            '-c 4 '\\\n",
    "            '-a e '\\\n",
    "            '-f e '\\\n",
    "            '--no_memory_check '\\\n",
    "            f'-i {aln_phy}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!phy_cmd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phyml -d aa -m LG -b -4 -v 0.0 -c 4 -a e -f e --no_memory_check -i ../data/tree/P9/P9.phy\n"
     ]
    }
   ],
   "source": [
    "print(phy_cmd)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
